beacon-chain/blockchain/forkchoice_update_execution.go

```diff
@@ -0,0 +1,61 @@
+package blockchain
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/v3/beacon-chain/state"
+	"github.com/prysmaticlabs/prysm/v3/consensus-types/interfaces"
+)
+
+func (s *Service) isNewProposer() bool {
+	_, _, ok := s.cfg.ProposerSlotIndexCache.GetProposerPayloadIDs(s.CurrentSlot()+1, [32]byte{} /* root */)
+	return ok
+}
+
+func (s *Service) isNewHead(r [32]byte) bool {
+	s.headLock.RLock()
+	defer s.headLock.RUnlock()
+	return r == [32]byte{} || r != s.headRoot()
+}
+
+func (s *Service) getHeadStateAndBlock(ctx context.Context, r [32]byte) (state.BeaconState, interfaces.SignedBeaconBlock, error) {
```

Comment:
This function name is very misleading as it does not get the head state and block, but rather any state and block with the passed block root `r`.